diff --git a/core/lib/Drupal/Component/Utility/Number.php b/core/lib/Drupal/Component/Utility/Number.php
index 21bbfe732f4..567f999651f 100644
--- a/core/lib/Drupal/Component/Utility/Number.php
+++ b/core/lib/Drupal/Component/Utility/Number.php
@@ -9,6 +9,71 @@
  */
 class Number {
 
+  /**
+   * Normalizes a numeric value to a lossless, easily parseable numeric string.
+   *
+   * The normalized value is a string suitable for use with libraries such as
+   * BCMath.
+   *
+   * @param int|float|string $number
+   *   The value to normalize. If its a string, it must be formatted as an
+   *   integer or float. Float with a higher number of significant decimals
+   *   than the precision value from the PHP runtime configuration
+   *   [default: 14] will lose the additional precision as PHP does not
+   *   guarantee that precision value.
+   *
+   * @see https://secure.php.net/manual/en/book.bc.php
+   *
+   * @return string
+   *   The normalized numeric string.
+   */
+  public static function normalize($number): string {
+    if (is_float($number)) {
+      // If the float has less significant decimals than the number we can
+      // guarantee, convert it to a string directly.
+      if (preg_match(sprintf('/^\d+\.\d{1,%d}$/', ini_get('precision')), (string) $number)) {
+        return (string) $number;
+      }
+      // For floats with more significant decimals than the number we can
+      // guarantee, discard those that are not guaranteed.
+      return rtrim(number_format($number, ini_get('precision'), '.', ''), '0');
+    }
+    return (string) $number;
+  }
+
+  /**
+   * Counts a number's significant decimals.
+   *
+   * @param int|float|string $number
+   *   The number whose decimals needed to be to counted. If this is a string,
+   *   it must be an integer or formatted as a float. Floats are limited to the
+   *   precision guaranteed by PHP (for example, 15).
+   *
+   * @return int
+   *   The number of significant decimal digits. Floats are limited to the
+   *   precision guaranteed by PHP (for example, 15). Numeric strings do not
+   *   suffer from system-specific limitations to float precision, so they
+   *   can contain more significant decimals.
+   *
+   *   @code Number::countSignificantDecimals(100.12345678901234567890) @endcode
+   *   returns 11 but
+   *   @code Number::countSignificantDecimals('100.12345678901234567890') @endcode
+   *   returns 20.
+   */
+  public static function countSignificantDecimals($number): int {
+    $number = is_numeric($number) ? static::normalize($number) : 0;
+
+    // If no decimal separator is encountered, there are 0 significant
+    // decimals.
+    if (!str_contains($number, '.')) {
+      return 0;
+    }
+
+    // If a decimal separator is encountered, count the number of significant
+    // decimal digits.
+    return strlen($number) - strrpos($number, '.') - 1;
+  }
+
   /**
    * Verifies that a number is a multiple of a given step.
    *
@@ -17,11 +82,20 @@ class Number {
    *
    * This is based on the number/range verification methods of webkit.
    *
-   * @param float $value
-   *   The value that needs to be checked.
-   * @param float $step
-   *   The step scale factor. Must be positive.
-   * @param float $offset
+   * Besides integers and floating numbers, we also support decimal numbers
+   * which are not stored in IEEE 754 format. In somewhat higher precisions for
+   * these numbers, the $step value cannot accurately represent the desired
+   * precision, when it is passed as a float. Passing it as a string bypasses
+   * this loss of precision and enables a correct calculation of the step
+   * validity.
+   *
+   * @param int|float|string $value
+   *   The value that needs to be checked. If this is a string, it must be
+   *   formatted as an integer or a float.
+   * @param int|float|string $step
+   *   The step scale factor. If this is a string, it must be formatted as an
+   *   integer or a float.
+   * @param float|null $offset
    *   (optional) An offset, to which the difference must be a multiple of the
    *   given step.
    *
@@ -31,7 +105,27 @@ class Number {
    * @see http://opensource.apple.com/source/WebCore/WebCore-1298/html/NumberInputType.cpp
    */
   public static function validStep($value, $step, $offset = 0.0) {
-    $double_value = (float) abs($value - $offset);
+    // Confirm the step is positive.
+    if ($step <= 0) {
+      return FALSE;
+    }
+
+    // Convert the value to a float so we can evaluate the precision later.
+    // Because subtracting the offset may change the value's precision, we only
+    // do so if it was set explicitly (is not null).
+    $float_value = (float) abs($value - $offset);
+
+    // The expected number significant decimals is dictated by the step.
+    $expected_significant_decimals = static::countSignificantDecimals($step) + 1;
+
+    // If the actual value has more significant decimals than expected, then it
+    // has a higher precision than desired, so it isn't divisible by the step.
+    $actual_significant_decimals = static::countSignificantDecimals($float_value);
+    if ($actual_significant_decimals > $expected_significant_decimals) {
+      return FALSE;
+    }
+
+    $float_value = (float) round($float_value, $expected_significant_decimals);
 
     // The fractional part of a double has 53 bits. The greatest number that
     // could be represented with that is 2^53. If the given value is even bigger
@@ -39,12 +133,13 @@ public static function validStep($value, $step, $offset = 0.0) {
     // remainder. Since that remainder can't even be represented with a single
     // precision float the following computation of the remainder makes no sense
     // and we can safely ignore it instead.
-    if ($double_value / pow(2.0, 53) > $step) {
+    if ($float_value / pow(2.0, 53) > $step) {
       return TRUE;
     }
 
+    $expected_float_value = (float) round($step * round($float_value / $step), $expected_significant_decimals);
     // Now compute that remainder of a division by $step.
-    $remainder = (float) abs($double_value - $step * round($double_value / $step));
+    $remainder = (float) abs($float_value - $expected_float_value);
 
     // $remainder is a double precision floating point number. Remainders that
     // can't be represented with single precision floats are acceptable. The
diff --git a/core/lib/Drupal/Core/Field/Plugin/Field/FieldType/DecimalItem.php b/core/lib/Drupal/Core/Field/Plugin/Field/FieldType/DecimalItem.php
index 6997c1bb8fa..629c8f3b8f2 100644
--- a/core/lib/Drupal/Core/Field/Plugin/Field/FieldType/DecimalItem.php
+++ b/core/lib/Drupal/Core/Field/Plugin/Field/FieldType/DecimalItem.php
@@ -74,7 +74,7 @@ public function storageSettingsForm(array &$form, FormStateInterface $form_state
       '#type' => 'number',
       '#title' => $this->t('Precision'),
       '#min' => 10,
-      '#max' => 32,
+      '#max' => max((int) ini_get('precision'), 10),
       '#default_value' => $settings['precision'],
       '#description' => $this->t('The total number of digits to store in the database, including those to the right of the decimal.'),
       '#disabled' => $has_data,
@@ -84,7 +84,7 @@ public function storageSettingsForm(array &$form, FormStateInterface $form_state
       '#type' => 'number',
       '#title' => $this->t('Scale', [], ['context' => 'decimal places']),
       '#min' => 0,
-      '#max' => 10,
+      '#max' => max((int) ini_get('precision'), 10),
       '#default_value' => $settings['scale'],
       '#description' => $this->t('The number of digits to the right of the decimal.'),
       '#disabled' => $has_data,
@@ -100,8 +100,10 @@ public function fieldSettingsForm(array $form, FormStateInterface $form_state) {
     $element = parent::fieldSettingsForm($form, $form_state);
     $settings = $this->getSettings();
 
-    $element['min']['#step'] = pow(0.1, $settings['scale']);
-    $element['max']['#step'] = pow(0.1, $settings['scale']);
+    // Convert to string, for consistent and lossless processing.
+    $step = number_format(pow(0.1, $settings['scale']), $settings['scale'], '.', '');
+    $element['min']['#step'] = $step;
+    $element['max']['#step'] = $step;
 
     return $element;
   }
diff --git a/core/lib/Drupal/Core/Field/Plugin/Field/FieldWidget/NumberWidget.php b/core/lib/Drupal/Core/Field/Plugin/Field/FieldWidget/NumberWidget.php
index e827bfd999e..227fae5d8ce 100644
--- a/core/lib/Drupal/Core/Field/Plugin/Field/FieldWidget/NumberWidget.php
+++ b/core/lib/Drupal/Core/Field/Plugin/Field/FieldWidget/NumberWidget.php
@@ -79,7 +79,7 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
     // Set the step for floating point and decimal numbers.
     switch ($this->fieldDefinition->getType()) {
       case 'decimal':
-        $element['#step'] = pow(0.1, $field_settings['scale']);
+        $element['#step'] = number_format(pow(0.1, $field_settings['scale']), $field_settings['scale'], '.', '');
         break;
 
       case 'float':
diff --git a/core/modules/field/tests/src/Functional/Number/NumberFieldTest.php b/core/modules/field/tests/src/Functional/Number/NumberFieldTest.php
index eaa94b00d51..3e6f9276c09 100644
--- a/core/modules/field/tests/src/Functional/Number/NumberFieldTest.php
+++ b/core/modules/field/tests/src/Functional/Number/NumberFieldTest.php
@@ -48,11 +48,13 @@ protected function setUp(): void {
   public function testNumberDecimalField(): void {
     // Create a field with settings to validate.
     $field_name = $this->randomMachineName();
+    // Having precision = 10 and scale = 4 means 6 numbers before the decimal
+    // point and 4 after.
     FieldStorageConfig::create([
       'field_name' => $field_name,
       'entity_type' => 'entity_test',
       'type' => 'decimal',
-      'settings' => ['precision' => 8, 'scale' => 4],
+      'settings' => ['precision' => 10, 'scale' => 4],
     ])->save();
     FieldConfig::create([
       'field_name' => $field_name,
@@ -82,16 +84,29 @@ public function testNumberDecimalField(): void {
     $this->assertSession()->fieldValueEquals("{$field_name}[0][value]", '');
     $this->assertSession()->responseContains('placeholder="0.00"');
 
-    // Submit a signed decimal value within the allowed precision and scale.
-    $value = '-1234.5678';
-    $edit = [
-      "{$field_name}[0][value]" => $value,
+    // Submit decimal values within the allowed precision and scale.
+    $valid_entries = [
+      '-1234.5678',
+      '19999.0000',
+      '99999.0000',
+      '909888.96',
+      '909888.99',
+      '988999.0096',
+      '988999.0099',
     ];
-    $this->submitForm($edit, 'Save');
-    preg_match('|entity_test/manage/(\d+)|', $this->getUrl(), $match);
-    $id = $match[1];
-    $this->assertSession()->pageTextContains('entity_test ' . $id . ' has been created.');
-    $this->assertSession()->responseContains($value);
+
+    foreach ($valid_entries as $valid_entry) {
+      $this->drupalGet('entity_test/add');
+      $edit = [
+        "{$field_name}[0][value]" => $valid_entry,
+      ];
+      $this->submitForm($edit, 'Save');
+      preg_match('|entity_test/manage/(\d+)|', $this->getUrl(), $match);
+      $id = $match[1];
+      $this->assertSession()->pageTextContains('entity_test ' . $id . ' has been created.');
+      $this->assertSession()->responseContains($valid_entry);
+      $this->assertSession()->responseNotContains("{$field_name} is not a valid number.");
+    }
 
     // Try to create entries with more than one decimal separator; assert fail.
     $wrong_entries = [
diff --git a/core/tests/Drupal/Tests/Component/Utility/NumberTest.php b/core/tests/Drupal/Tests/Component/Utility/NumberTest.php
index 17482990004..7c5705603c9 100644
--- a/core/tests/Drupal/Tests/Component/Utility/NumberTest.php
+++ b/core/tests/Drupal/Tests/Component/Utility/NumberTest.php
@@ -24,9 +24,12 @@ class NumberTest extends TestCase {
    * @dataProvider providerTestValidStep
    * @covers ::validStep
    *
-   * @param numeric $value
+   * @dataProvider providerTestValidStep
+   * @covers ::validStep
+   *
+   * @param int|float|string $value
    *   The value argument for Number::validStep().
-   * @param numeric $step
+   * @param int|float|string $step
    *   The step argument for Number::validStep().
    * @param bool $expected
    *   Expected return value from Number::validStep().
@@ -74,23 +77,50 @@ public static function providerTestValidStep() {
       [42, 10.5, TRUE],
       [1, 1 / 3, TRUE],
       [-100, 100 / 7, TRUE],
-      [1000, -10, TRUE],
 
       // Valid and very small float steps.
+      [1936.5, 3e-8, TRUE],
       [1000.12345, 1e-10, TRUE],
       [3.9999999999999, 1e-13, TRUE],
 
       // Invalid integer steps.
       [100, 30, FALSE],
       [-10, 4, FALSE],
+      [1000, -10, FALSE],
 
       // Invalid float steps.
       [6, 5 / 7, FALSE],
       [10.3, 10.25, FALSE],
+      [1000, -10.0, FALSE],
 
       // Step mismatches very close to being valid.
       [70 + 9e-7, 10 + 9e-7, FALSE],
-      [1936.5, 3e-8, FALSE],
+
+      // These floats are valid, but might fail due to precision (rounding)
+      // errors that are inherent to floating point numbers.
+      [3333333, 0.01, TRUE],
+      [4444444, 0.01, TRUE],
+      [9990009888.96, 0.01, TRUE],
+      [9990009888.99, 0.01, TRUE],
+      [990088999.0096, 0.0001, TRUE],
+      [990088999.0099, 0.0001, TRUE],
+      [4031239412.52, 0.01, TRUE],
+      [90000000000.00, 0.01, TRUE],
+      [9999999999.99, 0.01, TRUE],
+      [-3.1933172, 0.0000001, TRUE],
+      [1109.87, 0.01, TRUE],
+      [70000000, 0.01, TRUE],
+      [70000000.00, 0.01, TRUE],
+      [13517282.20, 0.01, TRUE],
+      [13517282.21, 0.01, TRUE],
+
+      // Precision of the value is higher than that of the step.
+      [990088999.0099, 0.001, FALSE],
+      [990088999.0099, 0.01, FALSE],
+
+      // Ensure floats with more significant decimals that the guaranteed
+      // minimum are normalized.
+      [0.0000000000000900009, 0.0000000000000100001, TRUE],
     ];
   }
 
