diff --git a/gin_lb.api.php b/gin_lb.api.php
index 4a3c7ce..4773ff3 100644
--- a/gin_lb.api.php
+++ b/gin_lb.api.php
@@ -41,6 +41,25 @@ function hook_gin_lb_is_layout_builder_route_alter(&$gin_lb_is_layout_builder_ro
   }
 }
 
+/**
+ * Alter the template suggestion logic.
+ *
+ * Allows modules to overrule whether template suggestions should be added.
+ *
+ * @param boolean $add_suggestions
+ *   Boolean flag.
+ * @param array $variables
+ *   The variables array of a render item.
+ * @param string $hook
+ *   The theme hook.
+ */
+function hook_gin_lb_add_suggestions_alter(&$add_suggestions, $variables, $hook) {
+  $route_name = \Drupal::routeMatch()->getRouteName();
+  if ($route_name == 'layout_builder.add_block' && $add_suggestions) {
+    $add_suggestions = FALSE;
+  }
+}
+
 /**
  * @} End of "addtogroup hooks".
  */
diff --git a/gin_lb.module b/gin_lb.module
index 931acbc..75b51ac 100644
--- a/gin_lb.module
+++ b/gin_lb.module
@@ -167,6 +167,17 @@ function gin_lb_views_pre_render(ViewExecutable $view) {
   }
 }
 
+/**
+ * Implements hook_theme_registry_alter().
+ */
+function gin_lb_theme_registry_alter(&$theme_registry) {
+  foreach ($theme_registry as &$value) {
+    if (array_key_exists('variables', $value)) {
+      $value['variables']['gin_lb_theme_suggestions'] = NULL;
+    }
+  }
+}
+
 /**
  * Implements hook_theme_suggestions_alter().
  */
@@ -238,6 +249,10 @@ function gin_lb_theme_suggestions_alter(
     )) {
     $add_suggestions = TRUE;
   }
+
+  // Allow other modules to chime in.
+   \Drupal::moduleHandler()->alter('gin_lb_add_suggestions', $add_suggestions, $variables, $hook);
+
   if (isset($variables['element']['#gin_lb_form']) || $add_suggestions === TRUE) {
 
     // Fix form element suggestions when they are not implemented in the theme.
