diff --git a/core/lib/Drupal/Core/Plugin/Context/ContextHandler.php b/core/lib/Drupal/Core/Plugin/Context/ContextHandler.php
index 248cf7adbf..3914f7521c 100644
--- a/core/lib/Drupal/Core/Plugin/Context/ContextHandler.php
+++ b/core/lib/Drupal/Core/Plugin/Context/ContextHandler.php
@@ -104,8 +104,13 @@ public function applyContextMapping(ContextAwarePluginInterface $plugin, $contex
 
         // Pass the value to the plugin if there is one.
         if ($contexts[$context_id]->hasContextValue()) {
+          // Passed in context.
           $plugin->setContext($plugin_context_id, $contexts[$context_id]);
         }
+        elseif ($plugin_context->hasContextValue()) {
+          // Default context value.
+          $plugin->setContext($plugin_context_id, $plugin_context);
+        }
         elseif ($plugin_context_definition->isRequired()) {
           // Collect required contexts that exist but are missing a value.
           $missing_value[] = $plugin_context_id;
