diff --git a/core/modules/toolbar/tests/src/Functional/ToolbarAdminMenuTest.php b/core/modules/toolbar/tests/src/Functional/ToolbarAdminMenuTest.php
index 10af7ea5d94..4d3a851573d 100644
--- a/core/modules/toolbar/tests/src/Functional/ToolbarAdminMenuTest.php
+++ b/core/modules/toolbar/tests/src/Functional/ToolbarAdminMenuTest.php
@@ -91,6 +91,7 @@ protected function setUp(): void {
       'administer modules',
       'administer permissions',
       'administer users',
+      'administer user roles',
       'access user profiles',
       'administer taxonomy',
       'administer languages',
diff --git a/core/modules/user/src/AccountForm.php b/core/modules/user/src/AccountForm.php
index c74619fddc7..21ef5bbe344 100644
--- a/core/modules/user/src/AccountForm.php
+++ b/core/modules/user/src/AccountForm.php
@@ -205,7 +205,7 @@ public function form(array $form, FormStateInterface $form_state) {
       '#title' => $this->t('Roles'),
       '#default_value' => (!$register ? $account->getRoles() : []),
       '#options' => $roles,
-      '#access' => $roles && $user->hasPermission('administer permissions'),
+      '#access' => $roles && $user->hasPermission('administer user roles'),
     ];
 
     // Special handling for the inevitable "Authenticated user" role.
@@ -214,6 +214,16 @@ public function form(array $form, FormStateInterface $form_state) {
       '#disabled' => TRUE,
     ];
 
+    // Prevent non-admin users from assigning the admin role to anybody.
+    // Remove, once https://www.drupal.org/project/drupal/issues/2846365
+    // is resolved
+    if (!$user->hasPermission('administer permissions')) {
+      $form['account']['roles']['administrator'] = [
+        '#default_value' => FALSE,
+        '#disabled' => TRUE,
+      ];
+    }
+
     $form['account']['notify'] = [
       '#type' => 'checkbox',
       '#title' => $this->t('Notify user of new account'),
diff --git a/core/modules/user/tests/src/Functional/UserRolesAssignmentTest.php b/core/modules/user/tests/src/Functional/UserRolesAssignmentTest.php
index 3a842409efe..ebae42bb458 100644
--- a/core/modules/user/tests/src/Functional/UserRolesAssignmentTest.php
+++ b/core/modules/user/tests/src/Functional/UserRolesAssignmentTest.php
@@ -17,8 +17,8 @@ class UserRolesAssignmentTest extends BrowserTestBase {
   protected function setUp(): void {
     parent::setUp();
     $admin_user = $this->drupalCreateUser([
-      'administer permissions',
       'administer users',
+      'administer user roles',
     ]);
     $this->drupalLogin($admin_user);
   }
diff --git a/core/modules/user/user.permissions.yml b/core/modules/user/user.permissions.yml
index c32c7e7b6f4..3aaad1f9442 100644
--- a/core/modules/user/user.permissions.yml
+++ b/core/modules/user/user.permissions.yml
@@ -9,6 +9,10 @@ administer users:
   title: 'Administer users'
   description: 'Manage all user accounts. This includes editing all user information, changes of email addresses and passwords, issuing emails to users and blocking and deleting user accounts.'
   restrict access: true
+administer user roles:
+  title: 'Administer users roles'
+  description: 'Add or remove roles on user accounts. Note: to administer user roles, you also need the permission to administer users.'
+  restrict access: true
 access user profiles:
   title: 'View user information'
 view user email addresses:
