diff --git a/webp.module b/webp.module
index de27098..7f3aeea 100644
--- a/webp.module
+++ b/webp.module
@@ -47,6 +47,9 @@ function webp_preprocess_responsive_image(&$variables) {
         /* @var \Drupal\webp\Webp $webp */
         $webp = \Drupal::service('webp.webp');
         $webp_srcset = $webp->getWebpSrcset($srcset_orig);
+        if ($webp_srcset && !file_exists($webp_srcset)) {
+          $webp->createWebpCopy($webp_srcset);
+        }
 
         // Create a new source pointing to the webp URL.
         $webp_source = clone $source;
